---
title: "Annotating, Filtering and Prioritising"
author: "Mark Dunning"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
---

# Outline

Imagine that we've called a set of somatic variants from our paired tumour and normal sample. From the previous section we can define a set of filters that can give us confidence that our variants are legitimate biological differences, and not technical noise. In this session we will go to the next step and see how we can gain biological intuition from our calls. 

- Using `annovar` to retrieve and annotate using pre-defined databases
- Using `annovar` to create a human-readable and searchable table of annotation results
- Recent advances in R ("`dplyr`") that allow us to filter and subset our data more efficiently
- How we can prioritise our results, and write pipelines for doing so

The calls we are going to be using is `HCC1143_vs_HCC1143_BL.flagged.muts.vcf`, which should be in the `Day3` directory

# Working Directory

```{}
/home/participant/Course_Materials/Day3
```


## Why annotate

- Can have huge list of variants after SNV-calling
    + order of millions even after filtering
- Not all will have consequences for the individual
- Need to functionally annotate biological and biochemical function
- Determine which variants have been previously-identified
    + in previous studies
    + amongst healthy / diseased individuals
    
## Introduction to annovar

There are many annotation resources available, but we have chosen to demonstrate `annovar`. There is an overhead associated with downloading the required annotation files, and a lot of disk space is consumed. However, you will hopefully agree that once configured it is relatively easy to use. `annovar` is a suite of perl scripts that can perform a variety of annotation and filtering tasks. It can also collate the results from annotating against different databases into one convenient table.

***Important:- annovar license***

`annovar` is free for academic use. However, you need to supply a valid academic email address in order to download the software. If you wish to use `annovar` after the course. You will have to fill out the form on the [annovar](http://annovar.openbioinformatics.org/en/latest/user-guide/download/) website. 


`annovar` provides "gene-based", region-based" and "filter-based" annotation and relies on *pre-downloaded* annotation files so all queries are done offline. Files can be downloaded from the UCSC genome browser or from the annovar website. Some of these files can be quite large, but fortunately we only have to download them once. For the course, we have downloaded some of these files to the directory `/home/participant/Course_Materials/software/annovar/humandb`. Each annotation file has a genome version associated with it. For our purposes we are going to be using `hg19`. 

For your reference the script to download all the resources used in this session is

`/home/participant/Course_Materials/software/annovar/annovar_commands.sh`


## Preparing files for annovar

Some of the commands in `annovar` require a specific input format. Fortunately, they have provided a script that can convert from popular file formats to their own format. 

We convert our `.vcf` file using a script from annovar `convert2annovar.pl`

```{}
../software/annovar/convert2annovar.pl -format vcf4old  HCC1143_vs_HCC1143_BL.flagged.muts.vcf > HCC1143_vs_HCC1143_BL.flagged.muts.avinput
```

```{r echo=FALSE,eval=FALSE}
system("../software/annovar/convert2annovar.pl -format vcf4old  HCC1143_vs_HCC1143_BL.flagged.muts.vcf > HCC1143_vs_HCC1143_BL.flagged.muts.avinput")
```


```{}
-format: vcfold, the input is in vcf format (old-style)
-includeinfo: Include the INFO fields from the vcf file
>: redirect to the file of our choosing
```


```{r echo=FALSE,comment=NA} 
cat(system("head -n2 HCC1143_vs_HCC1143_BL.flagged.muts.avinput",intern=TRUE),sep="\n")
```

The first five columns are compulsory and contain

- Chromosome
- Start Position
- End Position
- Reference base(s)
- Observed base(s)

All other columns are optional. These columns are not used in the annotation process, but will get carried-along with any annotation and filtering we perform if we wish. 

### Region-based Annotation

Gene-based annotation is the default operation in `annovar` and is used to identify variants that cause protein-coding changes. Various gene-definition systems are supported.

Database  | Keyword
------------- | -------------
RefSeq  | refGene
UCSC    | knownGene
Ensembl | ensGene
Gencode | wgEncodeGencodeBasicV19


The following command will annotate the file we have just prepared against the reference file located in `../software/annovar/humandb/hg19_refGene.txt`. We would normally have to specify an annotation mode and database keyword. However, region-based annotation and refGene are the defaults for the annotation script.

```{}
../software/annovar/annotate_variation.pl -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb
```

```{r echo=FALSE,eval=FALSE}
system("../software/annovar/annotate_variation.pl -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb")
```


Two files are created:

-`HCC1143_vs_HCC1143_BL.flagged.muts.avinput.variant_function`

This file lists ***all*** variants in the input file are pre-pends two columns. The first states whether a given variant is within an exonic or not and gives the relevant gene name. 

```{r echo=FALSE,comment=NA}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.variant_function",intern=TRUE),sep="\n")
```

The second file is

- `HCC1143_vs_HCC1143_BL.flagged.muts.avinput.exonic_variant_function` 

```{r echo=FALSE,comment=NA}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.exonic_variant_function",intern=TRUE),sep="\n")
```

This file lists for the ***coding variants only*** whether the stated base-change will lead to a change in protein or not. Full definitions of all the terms is on the [annovar website](http://annovar.openbioinformatics.org/en/latest/user-guide/gene/)


### Region-based annotation

Regioin-based annotation is used when we want to see if our variants fall within pre-defined genomic regions of interest. Here we list some examples along with the "*keyword*" to be used when downloading and annotating against the database. More comprehensive details on the types of region-based filtering that is supported can be found on the [http://annovar.openbioinformatics.org/en/latest/user-guide/region/](annovar website). 


Database  | Keyword 
------------- | -------------
[Cytoband](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#identify-cytogenetic-band)  | cytoband 
[microRNA and snoRNA](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#identify-variants-disrupting-micrornas-and-snornas) | wgRna 
[Conserved Regions](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#conserved-genomic-elements-annotation)  | phastConsElements46way
[Transcription Factor Binding Sites](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#transcription-factor-binding-site) | tfbsConsSites
[Published GWAS](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#identify-variants-reported-in-previously-published-gwas) | gwasCatalog
[Database of Genomic Variants](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#identify-previously-reported-structural-variants) | dgvMerged
[Segmental Duplications](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#identify-variants-located-in-segmental-duplications) | genomicSuperDups

Annotation against regions identified in [ENCODE](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#identify-variants-in-encode-annotated-regions) project is also described in its own section on the annovar website. Alternatively, you can specify your own regions of interest in [.bed](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#identifying-variants-in-regions-specified-in-bed-files) or [.gff](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#annotating-custom-made-databases-conforming-to-gff3-generic-feature-format-version-3) format.

The simplest form of region annotation is probably when we want to look at which "cytoband" each variant belongs to. This uses a reference file that contains the coordinates for each cytoband in a given genome version. 

```{}
head ../software/annovar/humandb/hg19_cytoBand.txt
```

```{r echo=FALSE,comment=NA}
cat(system("head ../software/annovar/humandb/hg19_cytoBand.txt",intern=TRUE),sep="\n")
```

If didn't already have such a file it could be downloaded using the following. The `cytoBand` keyword (from the table above) is used along with the build version `hg19`.

```{}
../software/annovar/annotate_variation.pl -buildver hg19 -downdb cytoBand ../software/annovar/humandb/
```



To perform the annotation, we use need to specify `-regionanno`, to perform region-annotation and set the `dbtype` to `cytoBand`. 

```{}
../software/annovar/annotate_variation.pl -regionanno -build hg19 -dbtype cytoBand  HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/
``` 

```{r echo=FALSE,eval=FALSE}
system("../software/annovar/annotate_variation.pl -regionanno -build hg19 -dbtype cytoBand  HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/")
```


This should result in a file being created with `.cytoBand` added as a file extension to your input file (e.g. `HCC1143_vs_HCC1143_BL.flagged.muts.avinput.cytoBand`). An extra column has been added to indicate which chromosome band each variant is located on. In the event of a variant spanning multiple bands (obviously not in this case), the multiple bands would be listed here.

```{}
head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.cytoBand
```

```{r echo=FALSE,comment=NA}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_cytoBand",intern=TRUE),sep="\n")
```

*******
*******
*******

Exercise

- Select any of the databases from the table that you think might be relevant to your study, and annotate the example SNV calls
    + all the database files should already be downloaded.
    + try the GWAS catalogue if you are not sure    

*******
*******
*******


### Filter-based annotation

Unlike the annotation performed in the previous section, which was based purely on the genomic location, *filter-based* annotation also takes the base-change into account. One can also use the observed frequency in the database population as a means to filter.

As before, many types of database are available. An additional complication compared to the region filters from the previous section is that several version numbers exist for each database. In this table, we list the *latest* version available as of July 2016.

Database  | Keyword 
------------- | -------------
[dbSNP](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#dbsnp-annotations)  | snp138
[1000 Genomes](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#1000-genomes-project-2015-aug-annotations) | 1000g2015aug
[Exome Aggregation Consortium](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#exac-annotations) | exac03
[Exome Sequencing Project](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#esp-exome-sequencing-project-annotations) | esp6500si_all 
[Complete Genomics Individuals](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#cg-complete-genomics-frequency-annotations) | cg69
[ClinVar](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#clinvar-annotations) | clinvar_20160302
[COSMIC](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#cosmic-annotations) | cosmic70
[ICGC](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#icgc-annotations) | icgc21
[NCI60](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#nci60-annotations) | nci60

As we will see, filter-based annotation is much the same as region- and gene-based. However, when downloading for the first time we have to add an additional `-webfrom annovar` argument into the command as the filter-based files are provided by the annovar website. For the purposes of this session, all the files have already been provided. Here is how we would download the cosmic data.

```{}
../software/annovar/annotate_variation.pl -buildver hg19 -downdb -webfrom annovar cosmic70 ../software/annovar/humandb/
```

It seems natural to use the latest version of a database for the most-comprehensive annotation. However, the annovar website suggests that we be cautious with regards to [dbSNP](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#dbsnp-annotations). As dbSNP has been expanded to include variants from other large-scale projects (such as 1000 genomes) it has become somewhat "contaminated" by variants with very low frequency or with clinical relevance. Thus, ***dbSNP129*** is regarded as being the last clean version of the database.


To filter against dbSNP 129 we select the `filter`  to `.annotate_variation.pl` with a `dbtype` of `snp129`
```{}
../software/annovar/annotate_variation.pl -filter -buildver hg19 -dbtype snp129  HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/
```

```{r echo=FALSE,eval=FALSE}
system("../software/annovar/annotate_variation.pl -filter -buildver hg19 -dbtype snp129  HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/")
```


This time we get two files created; `HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_snp129_dropped` and `HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_snp129_filtered`. The difference being that the `.._dropped` file contains details of all your variants that ***are*** in the database

```{}
head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_snp129_dropped
```

```{r echo=FALSE,comment=NA,eval=FALSE}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_snp129_dropped",intern=TRUE),sep="\n")
```

Thus, the `.._filtered` file contains all the variants ***not found*** in the database. The implication being that we want to remove common variants found amongst healthy individuals from our list. There is nothing to stop us using multiple versions of dbSNP in the filtering as we will see later-on.

We can perform annotation against 1000 genomes dataset with the following:-

```{}
../software/annovar/annotate_variation.pl -filter -dbtype 1000g2015aug_all -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/
```

```{r echo=FALSE,eval=FALSE}
system("../software/annovar/annotate_variation.pl -filter -dbtype 1000g2015aug_all -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/")
```


As before, `..._dropped` and `...filtered` files are produce. Again the `.._dropped` file contains variants that were found in the database. However, rather than printing and ID for the variant, it prints the minor allele frequency that is observed for that variant.

```{}
head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_ALL.sites.2015_08_dropped
```

```{r echo=FALSE,comment=NA,eval=FALSE}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_ALL.sites.2015_08_dropped",intern=TRUE),sep="\n")
```

The commands above annotated against variants derived from *all** indivduals in the study. If we have individuals of a different ethnicity, we probably want to use a more-specific cohort. In this case, we annotate against frequencies observed in *european* individuals.

```{}
../software/annovar/annotate_variation.pl -filter -dbtype 1000g2015aug_eur -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/
```

## Healthy or disease individuals?

How you choose to interpret the filtering results depends on whether you are interested in variants that are exclusive to healthy or diseased individuals.

Both dbSNP and 1000 genomes report frequencies amongst healthy individuals. The "complete genomics" (`cg69`) set also provides variant frequencies for 69 healthy subjects (including 46 unrelated subjects). Particularly useful for those with exome data is the ESP (Exome Sequencing Project) which aims to identify variants in exonic regions from over 6000 individuals; including however some subjects with disease. 


`annovar` can also annotate against variants that are relevant to diseases such as cancer. For example, there is the ***COSMIC*** (Catalogue of Somatic Mutations in Cancer).

```{}
../software/annovar/annotate_variation.pl -filter -dbtype cosmic70 -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/
```

```{r echo=FALSE,eval=FALSE}
system("../software/annovar/annotate_variation.pl -filter -dbtype cosmic70 -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/")
```


Again, the variants found in the database are printed in the `_dropped` file along with details of the mutation in **COSMIC** that they match.

```{r echo=FALSE,comment=NA}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_cosmic70_dropped",intern=TRUE),sep="\n")
```

This is slightly contrary to the 1000 genomes and dbSNP filtering we did previously where we wanted to retain variants that were not found in a database. Fortunately in the next section we will look at any approach where we can get results for all our variants in a single table that we can develop our own filtering criteria for.

### Predicting consequence

`annovar` also has a wide array of filters that can help us to [predict the consequence](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#ljb42-dbnsfp-non-synonymous-variants-annotation) of a particular mutation. These include predictions from methods such as [Sifting Intolerant From Tolerant / SIFT](http://sift.jcvi.org/), [Polyphen](http://genetics.bwh.harvard.edu/pph/) , [MutationTaster](http://mutationtaster.org/) and [FATHMM](http://fathmm.biocompute.org.uk/). 

The keyword `dbsnfp30a` will annotate against all `annovar`s collection of prediction databases in one go. However, for the purposes of the practical we will not be running this as it takes a long time.

Each prediction method can however be run separately. If a particular variant is found in the database a score will be returned along with some prediction of affect on protein function. A summary of the output of some of these methods is shown below. Details of others are on the annovar website.


Method  | Keyword | Prediction
------------- | ------------ | -----
SIFT  | ljb23_sift | D: Deleterious (sift <= 0.05); T: tolerated (sift >0.05)
MutationTaster  | ljb23_mt | A (disease causing automatic); D (disease causing); N (polymorphism); P (polymorphism automatic)
FatHMM | ljb23_fathmm | D: Deleterious; T: Tolerated

## Combining multiple annotations and filters

Up to this point, we have amassed quite a collection of tables of our set of variants annotated / filtered against various databases. A particularly appealing feature of `annovar` is that we can define protocols to combine multiple filters and collect the results in the same table. 

The `-protocol` argument is used to list what types of annotation / filtering you want to be applied using the same keywrods as previously. The `-operation` argument is used to specify whether each annotation is a gene (`g`), region (`r`) or filter (`f`) -based annotation. You need to take care to provide the same number of names in both the `-protocol` and `-operation` arguments. When you run the following command you should see that each filter is applied in turn.

```{}
../software/annovar/table_annovar.pl HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/ -buildver hg19 -out HCC1143_vs_HCC1143_BL.flagged.muts.annovar -remove -protocol refGene,cytoBand,gwasCatalog,genomicSuperDups,dgvMerged,snp129,1000g2015aug_all,esp6500si_all,clinvar_20160302,cosmic70,nci60,ljb23_sift,ljb23_mt,ljb23_fathmm -operation g,r,r,r,r,f,f,f,f,f,f,f,f,f -nastring NA -csvout
```

```{r table_annovar, echo=FALSE,eval=FALSE}
system("../software/annovar/table_annovar.pl HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/ -buildver hg19 -out HCC1143_vs_HCC1143_BL.flagged.muts.annovar -remove -protocol refGene,cytoBand,gwasCatalog,genomicSuperDups,dgvMerged,snp129,1000g2015aug_all,esp6500si_all,clinvar_20160302,cosmic70,nci60,ljb23_sift,ljb23_mt,ljb23_fathmm -operation g,r,r,r,r,f,f,f,f,f,f,f,f,f -nastring NA -csvout")
```


The file `HCC1143_vs_HCC1143_BL.flagged.muts.annovar.hg19_multianno.csv` is created. By specifying the `-remove` argument in the command, any tables that are created as part of the process are removed. This `.csv` file can now be viewed in Excel, or LibreOffice in our case (right-click on the file)

Having opened this file into a spreadsheet program we can of course filter and sort the data as we see fit. However, due to the sheer number of columns we have generated this may prove a bit unwieldy. Therefore we are going to introduce an R package that will make the interrogation and comprehension of such a table more intuitive.

# Using R to interrogate and filter an annotation table

## Introducing dplyr

`dplyr` is a very powerful package for manipulating data in tables and is part of the *tidy-verse* collection of packages that provide a seamless and consistent interface to exploring and analysing data.

![](images/data-cycle.png)

The process of analysing a dataset is rarely a linear one, and will generate scores of intermediate plots and tables along the way that don't make their way to a final publication. Packages such as `dplyr` and `ggplot2` (for plotting) provide an interface whereby the effort to produce meaningful summaries and graphics of our data is minimised. This allows us to worry more about the interpretation and what questions we should ask, rather than writing code.

In the context of filtering variants this is useful, as we may have to re-visit our criteria many times before settling on a set of rules.

The [cheatsheet](https://www.rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf) is highly recommended and summarises all the functions within the package. We won't have time to cover everything here today.

For more information on `dplyr` see our one-day course [Data Manipulation and Visualisation using R](http://bioinformatics-core-shared-training.github.io/r-intermediate/)

The first step is to load the `dplyr` library. We will also need `VariantAnnotation` for this workflow.

```{r message=FALSE}
library(VariantAnnotation)
library(dplyr)
```

An incredibly useful but simple function is `tbl_df` which converts the standard R `data.frame` into a special `dplyr` one that has a nicer standard display. Rather than filling the whole R console we can see a snapshot of the data frame and summaries of the data contained in each column


```{r}
anno <- read.csv("HCC1143_vs_HCC1143_BL.flagged.muts.annovar.hg19_multianno.csv")
anno <- tbl_df(anno)
anno
```

`dplyr` introduces a series of verbs, each of which is capable of performing a specific operation on a data frame in an efficient manner. The syntax of the verbs is very consistent. Not only does this allow us to easily switch between using different operations, but we can chain the operations together as we will see later.

### select

`select` is the first of the `dplyr` verbs that can be used to pick certain columns from the table

```{r}
select(anno, Func.refGene)
```

No great advance in our capabilities yet, as we already know how to select a variable from a table using the `$` operator

```{r eval=FALSE}
anno$Func.refGene
```

However, the syntax to select multiple columns is a lot more concise. Mutiple columns can be selected by separating their names with a comma

Compare:- 

```{r}
select(anno, Func.refGene,ExonicFunc.refGene)
```

to

```{r eval=FALSE}
anno[,c("Func.refGene","ExonicFunc.refGene")]
```

Non R users probably find these commands a bit obtuse

- why the `[ ]`?
- what is `c`?
- need to remember the row and column index
- `[,...]` means display all rows

However, `select` is a lot more intuitive; easier to convert to the sentance “Select the Func.refGene and ExonicFunc.refGene columns from patients”

We can select a range of columns with a `:`

```{r}
select(anno, cytoBand:X1000g2015aug_all)
```

Or suppress columns by putting with a `-` in front of their name

```{r}
select(anno, -End)
```

Combinations of the above are possible

```{r}
select(anno, cytoBand:X1000g2015aug_all,-snp129,-AAChange.refGene)
```

Some convenient functions are available so we don't need to specify the column names in full

```{r}
select(anno, starts_with("CLN"))
select(anno, contains("1000g"))
```

*******
Exercise

- Display the columns `Func.refGene` to `AAChange.refGene` 
    + you could try the convenience functions `contains` `ends_with` to do this, or using the `:` notation
- Now remove the column `GeneDetail.refGene`
- Append the chromosome start and end position columns

```{r echo=FALSE}
select(anno, Chr:End, contains("refGene"),-GeneDetail.refGene)
```

******

We should note at this point, that we have not actually changed the data frame at any point; just created different views of the data. 

```{r}
anno
```


### `filter`

A particularly useful operation for us is going to be `filter`. Again, there is nothing here that we couldn't do with "base R", but it will allow us to create more readable code.

The syntax is consistent with `select` and uses the usual selection techniques in R such as `==`.

Selecting variants on chromsome 1

```{r}
filter(anno, Chr ==1)
```

Selecting *exonic* variants
```{r}
filter(anno, Func.refGene == "exonic")
```

Multiple conditions can be separated by a `,`. This is the same as using `&` in base R to combine various tests.

"*Exonic variants on chromosome 1 that are no synonymous or unknown*"
```{r}
filter(anno, Func.refGene == "exonic", Chr == 1, ExonicFunc.refGene != "synonymous SNV", ExonicFunc.refGene != "unknown")
```

Partial matches can be performed using `grepl`. We often use `grep` to find the indices where a particular string is found amongst a vector. However, in order to do filtering `dplyr` requires a logical vector the same length as the number of rows. 

```{r}
grep("breast",anno$cosmic70)
grepl("breast",anno$cosmic70)[1:10]
```


```{r}
filter(anno, grepl("breast",cosmic70))
```

During annovar's filtering we get `NA` entries whenever a particular variant is not found in a database. To filter for these `NA` values we use the function `is.na`.

e.g. Variants that are found in dbSNP
```{r}
filter(anno, is.na(snp129))
```

Variants that are found in GWAS. Were we use a `!` to reverse the result of `is.na`; `TRUE` -> `FALSE` and `FALSE` -> `TRUE`

```{r}
filter(anno, !is.na(gwasCatalog))
```

Variants that not found in 1000 genomes, ***or*** have a maf frequency less than 0.05. Here we use the operation `|` to combine different logical tests.

```{r}
filter(anno, is.na(X1000g2015aug_all) | X1000g2015aug_all < 0.05)
```



*******
Exercise

- Find which variants are not in segmental duplications
- Find all variants that are in TP53 and exonic
- Find variants with greater than 0.1 frequency in the Exome Sequencing Project
    + column `esp6500si_all`

*******

## Combining commands with "pipes"

A slighty annoyance with the exercise above is that once we've filtered for low frequency esp variants, we don't actually see the `esp6500si_all` column appearing in the output:-

```{r}
filter(anno, esp6500si_all > 0.1)
```

Given what we have done so far, it would seem natural to follow this `filter` statement with a `select` to view the columns we are interested in

It is quite common to nest commands in R into a single line

```{r eval=FALSE}
anno <- tbl_df(read.csv("HCC1143_vs_HCC1143_BL.flagged.muts.annovar.hg19_multianno.csv"))
```

which we read as 

> apply the function `tbl_df` to the result of reading the file `HCC1143_vs_HCC1143_BL.flagged.muts.annovar.hg19_multianno.csv`

The following line of R code is perfectly valid and does exactly what we want. However, reading the code back to ourselves is a bit clunky.

```{r}
select(filter(anno, esp6500si_all > 0.1), Chr:Alt, esp6500si_all)
```

> select from variants with more than 0.1 frequency in the exome sequencing project, the columns Chr to Alt and esp6500si_all

or we can split over multiple lines
```{r}
select(filter(anno, esp6500si_all > 0.1),
       Chr:Alt, esp6500si_all)
```


We always have to work out what the first statement was, and work forwards from that.
 - the arguments to a function are far-away from the function call


***piping*** is a familiar concept in unix where the output of one command is used as input to the following line. The `dplyr` package allows such operations to be performed in R via use of the ` %>% ` operation from the `magrittr` package

- Shortcut in RStudio: CTRL + SHIFT + M, CMD + SHIFT + M (Mac)

`dplyr` functions are designed to take a data frame as their first argument, and return a data frame as an output. This makes them ideal to use with a pipe.

The code from before can be rewritten as follows, making it easier to see the order in which the operations were performed.

```{r}
filter(anno, esp6500si_all > 0.1) %>% 
  select(Chr:Alt, cosmic70)
```


*******
*******
*******

Exercise

- Find which variants are present in the GWAS catalog
- Use `select` to print the chromsome, start, end, Gene name and 1000 genomes frequency and GWAS catalogue details for these variants
- Use a "pipe" command to write your answer

```{r echo=FALSE}
filter(anno, !is.na(gwasCatalog)) %>% 
  select(Chr:End, Gene.refGene, X1000g2015aug_all,gwasCatalog)
```


- Explain in words what the following chain of commands is doing

```{r}
filter(anno, Func.refGene=="exonic") %>% 
  filter(!is.na(genomicSuperDups)) %>% 
  filter(is.na(snp129)) %>% 
  filter(X1000g2015aug_all < 0.05 | is.na(X1000g2015aug_all)) %>% 
  select(Chr:ExonicFunc.refGene,-GeneDetail.refGene,-Func.refGene,X1000g2015aug_all)
```

*******
*******
*******


The `arrange` function can be used to order rows. For example ordering according to 1000 genomes frequency after the series of commands above


```{r}
  arrange(anno, X1000g2015aug_all)
```

The useful `desc` function can be used to reverse the ordering from highest to lowest

```{r}
  arrange(anno, desc(X1000g2015aug_all))
```


## Adding extra information to the table

The `mutate` operation allows us to compute and append one or more new columns to a data frame. We can specify a new column name, or overwrite an existing one. As above, we do not change the data frame so need to create a new variable if we want the changes to persist.

For example, we might want to add the "chr" prefix to all our chromosome names. In order to perform this operation we need to specify a function that will take a vector as an input and return as vector of the same length as an output. In other words, a function that will transform each item in a vector and return a vector of transformed values.

If you haven't seen the paste function before, it is used to join character vectors together with the option to separate by a fixed character. The default separater is a single space " ". However we don't want any separater in this case so specify the empty string "". Alternatively the convenient function paste0 can be used.

```{r}
mutate(anno, Chr = paste("Chr", Chr,sep=""))
mutate(anno, Chr = paste0("Chr", Chr))

```

```{r}
anno <- mutate(anno, Chr = paste("Chr", Chr,sep=""))
```

To aid us selecting variants for follow-up, we need to know which ones have passed filters. We can first import the `.vcf` file as we saw yesterday.

```{r}
vcf <- readVcf("HCC1143_vs_HCC1143_BL.flagged.muts.vcf","hg19")
```

Recall that the `FILTER` flag is stored in the metadata of the variant ranges object `rowRanges`. We can add this to our existing annotation using `mutate`.

```{r}
anno.extra <- mutate(anno, FILTER = mcols(rowRanges(vcf))$FILTER)  
anno.extra
```



## Putting it together..

We now have all the `dplyr` operations in place to create a filtering pipeline. Firstly, probably want to use only those variants that passed the filters employed by caveman.

```{r}
filter(anno.extra, FILTER=="PASS")
```

Now we select

- Variants that pass caveman filters
- Variants not in segmental duplications
- Variants less than 0.05 frequency in 1000 genomes data
- Variants less than 0.05 frequencing in Exome Sequencing Project data
- Exonic variants
- Variants that cause a change in amino acid

```{r}
filter(anno.extra, FILTER=="PASS") %>% 
  filter(is.na(genomicSuperDups)) %>% 
  filter(is.na(X1000g2015aug_all) | X1000g2015aug_all < 0.05) %>% 
  filter(is.na(esp6500si_all) | esp6500si_all < 0.05) %>% 
  filter(Func.refGene == "exonic") %>% 
  filter(ExonicFunc.refGene != "synonymous SNV")
```

We can further reduce the set to those that SIFT considers to be deleterious

```{r}
filter(anno.extra, FILTER=="PASS") %>% 
  filter(is.na(genomicSuperDups)) %>% 
  filter(is.na(X1000g2015aug_all) | X1000g2015aug_all < 0.05) %>% 
  filter(is.na(esp6500si_all) | esp6500si_all < 0.05) %>% 
  filter(Func.refGene == "exonic") %>% 
  filter(ExonicFunc.refGene != "synonymous SNV") %>% 
  filter(grepl("D", ljb23_sift)) %>% 
  select(Chr:Alt, Gene.refGene,ExonicFunc.refGene,cosmic70,ljb23_sift) 
```

Finally, write out to a `.csv` file

```{r}
filter(anno.extra, FILTER=="PASS") %>% 
  filter(is.na(genomicSuperDups)) %>% 
  filter(is.na(X1000g2015aug_all) | X1000g2015aug_all < 0.05) %>% 
  filter(is.na(esp6500si_all) | esp6500si_all < 0.05) %>% 
  filter(Func.refGene == "exonic") %>% 
  filter(ExonicFunc.refGene != "synonymous SNV") %>% 
  filter(grepl("D", ljb23_sift)) %>% 
  write.csv("HCC1143_vs_HCC1143_BL.flagged.muts.annovar.hg19_multianno.filtered.csv")
```

Depending on what your particular study is interested in, you may wish to omit or revise some of these steps.

