---
title: "Annotating and Filtering"
author: "Mark Dunning"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
---

# Outline

Imagine that we've called a set of somatic variants from our paired tumour and normal sample. From the previous section we can define a set of filters that can give us confidence that our variants are legitimate biological differences, and not technical noise. In this session we will go to the next step and see how we can gain biological intuition from our calls. 

- Using `annovar` to retrieve and annotate using pre-defined databases
- Using `annovar` to create a human-readable and searchable table of annotation results
- Recent advances in R ("`dplyr`") that allow us to filter and subset our data more efficiently
- How we can prioritise our results, and write pipelines for doing so

# Working Directory

```{}
/home/participant/Course_Materials/Day3
```


## Why annotate

- Can have huge list of variants after SNV-calling
    + order of millions even after filtering
- Not all will have consequences for the individual
- Need to functionally annotate biological and biochemical function
- Determine which variants have been previously-identified
    + in previous studies
    + amongst healthy / diseased individuals
    
## Introduction to annovar

There are many annotation resources available, but we have chosen to demonstrate `annovar` for ease of usage and compatibility across platforms. `annovar` is a suite of perl scripts that can perform a variety of annotation and filtering tasks. It can also collate the results from annotating against different databases into one convenient table.

***Important:- annovar license***

`annovar` is free for academic use. However, you need to supply a valid academic email address in order to download the software. If you wish to use `annovar` after the course. You will have to fill out the form on the [annovar](http://annovar.openbioinformatics.org/en/latest/user-guide/download/) website. 


`annovar` provides "gene-based", region-based" and "filter-based" annotation and relies on *pre-downloaded* annotation files so all queries are done offline. Files can be downloaded from the UCSC genome browser or from the annovar website. Some of these files can be quite large, but fortunately we only have to download them once. For the course, we have downloaded some of these files to the directory `/home/participant/Course_Materials/software/annovar/humandb`. Each annotation file has a genome version associated with it. For our purposes we are going to be using `hg19`. 

For your reference the script to download all the resources used in this session is

`/home/participant/Course_Materials/software/annovar/annovar_commands.sh`

```{}
ls -l ../software/annovar/humandb
```


## Preparing files for annovar

Some of the commands in `annovar` require a specific input format. Fortunately, they have provided a script that can convert from popular file formats to their own format. 

We convert our `.vcf` file using a script from annovar `convert2annovar.pl`

```{}
../software/annovar/convert2annovar.pl -format vcf4old -includeinfo HCC1143_vs_HCC1143_BL.flagged.muts.vcf > HCC1143_vs_HCC1143_BL.flagged.muts.avinput
```

```{}
-format: vcfold, the input is in vcf format
-includeinfo: Include the INFO fields from the vcf file
>: redirect to the file of our choosing
```


```{r echo=FALSE,comment=NA} 
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput",intern=TRUE),sep="\n")
```

The first five columns are compulsory and contain

- Chromosome
- Start Position
- End Position
- Reference base(s)
- Observed base(s)

All other columns are optional. They are not used in the annotation process, but will get carried-along with any annotation and filtering we perform. There is some duplication of information (e.g. position and base-change) but later will look at how to select particular columns of interest in R.

### Region-based Annotation

Gene-based annotation is the default operation in `annovar` and is used to identify variants that cause protein-coding changes. Various gene-definition systems are supported.

Database  | Keyword
------------- | -------------
RefSeq  | refGene
UCSC    | knownGene
Ensembl | ensGene
Gencode | wgEncodeGencodeBasicV19


The following command will annotate the file we have just prepared against the reference file located in `../software/annovar/humandb/hg19_refGene.txt`. We would normally have to specify an annotation mode and database keyword. However, region-based annotation and refGene are the defaults for the annotation script.

```{}
../software/annovar/annotate_variation.pl -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb
```

Two files are created:

-`HCC1143_vs_HCC1143_BL.flagged.muts.avinput.variant_function`

This file lists ***all*** variants in the input file are pre-pends two columns. The first states whether a given variant is within an exonic or not and gives the relevant gene name. 

```{r echo=FALSE,comment=NA}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.variant_function",intern=TRUE),sep="\n")
```

The second file is

- `HCC1143_vs_HCC1143_BL.flagged.muts.avinput.exonic_variant_function` 

```{r echo=FALSE,comment=NA}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.exonic_variant_function",intern=TRUE),sep="\n")
```

This file lists for the ***coding variants only*** whether the stated base-change will lead to a change in protein or not. Full definitions of all the terms is on the [annovar website](http://annovar.openbioinformatics.org/en/latest/user-guide/gene/)

If we wanted to annotate against UCSC gene definitions we need to add the `-dbtype` argument with the keyword `knownGene` (taken from the table above).

```{}
../software/annovar/annotate_variation.pl -buildver hg19 -dbtype knownGene HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb
```


### Region-based annotation

Regioin-based annotation is used when we want to see if our variants fall within pre-defined genomic regions of interest. Here we list some examples along with the "*keyword*" to be used when downloading and annotating against the database. More comprehensive details on the types of region-based filtering that is supported can be found on the [http://annovar.openbioinformatics.org/en/latest/user-guide/region/](annovar website). 


Database  | Keyword 
------------- | -------------
[Cytoband](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#identify-cytogenetic-band)  | cytoband 
[microRNA and snoRNA](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#identify-variants-disrupting-micrornas-and-snornas) | wgRna 
[Conserved Regions](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#conserved-genomic-elements-annotation)  | phastConsElements46way
[Transcription Factor Binding Sites](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#transcription-factor-binding-site) | tfbsConsSites
[Published GWAS](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#identify-variants-reported-in-previously-published-gwas) | gwasCatalog
[Database of Genomic Variants](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#identify-previously-reported-structural-variants) | dgvMerged
[Segmental Duplications](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#identify-variants-located-in-segmental-duplications) | genomicSuperDups

Annotation against regions identified in [ENCODE](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#identify-variants-in-encode-annotated-regions) project is also described in its own section on the annovar website. Alternatively, you can specify your own regions of interest in [.bed](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#identifying-variants-in-regions-specified-in-bed-files) or [.gff](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#annotating-custom-made-databases-conforming-to-gff3-generic-feature-format-version-3) format.

The simplest form of region annotation is probably when we want to look at which "cytoband" each variant belongs to. This uses a reference file that contains the coordinates for each cytoband in a given genome version. 

```{}
head ../software/annovar/humandb/hg19_cytoBand.txt
```

```{r echo=FALSE,comment=NA}
cat(system("head ../software/annovar/humandb/hg19_cytoBand.txt",intern=TRUE),sep="\n")
```

If didn't already have such a file it could be downloaded using the following. The `cytoBand` keyword (from the table above) is used along with the build version `hg19`.

```{}
../software/annovar/annotate_variation.pl -buildver hg19 -downdb cytoBand ../software/annovar/humandb/
```

To perform the annotation, we use need to specify `-regionanno`, to perform region-annotation and set the `dbtype` to `cytoBand`. 

```{}
../software/annovar/annotate_variation.pl -regionanno -build hg19 -dbtype cytoBand  HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/
``` 

This should result in a file being created with `.cytoBand` added as a file extension to your input file (e.g. `HCC1143_vs_HCC1143_BL.flagged.muts.avinput.cytoBand`). An extra column has been added to indicate which chromosome band each variant is located on. In the event of a variant spanning multiple bands (obviously not in this case), the multiple bands would be listed here.

```{}
head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.cytoBand
```

```{r echo=FALSE,comment=NA}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_cytoBand",intern=TRUE),sep="\n")
```

In another example, we could annotate against locations that have been reported in previous GWAS studies. The command is the same as the previous example, except we substitute `cytoBand` for the keyword for GWAS annotation (`gwasCatalog`)

```{}
../software/annovar/annotate_variation.pl -regionanno -build hg19 -dbtype gwasCatalog HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb
```

This time we get a file `HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_gwasCatalog`. The only entries in the file are those overlapping regions reported in GWAS studies

```{r echo=FALSE,comment=NA}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_gwasCatalog",intern=TRUE),sep="\n")
```

*******
Exercise

- Select any of the databases from the table that you think might be relevant to your study, and annotate the example SNV calls
    + all the database files should already be downloaded.
    
*******




### Filter-based annotation

Unlike the annotation performed in the previous section, which was based purely on the genomic location, *filter-based* annotation also takes the base-change into account. One can also use the observed frequency in the database population as a means to filter.

As before, many types of database are available. An additional complication compared to the region filters from the previous section is that several version numbers exist for each database. In this table, we list the *latest* version available as of July 2016.

Database  | Keyword 
------------- | -------------
[dbSNP](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#dbsnp-annotations)  | snp138
[1000 Genomes](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#1000-genomes-project-2015-aug-annotations) | 1000g2015aug
[Exome Aggregation Consortium](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#exac-annotations) | exac03
[Exome Sequencing Project](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#esp-exome-sequencing-project-annotations) | esp6500si_all 
[Complete Genomics Individuals](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#cg-complete-genomics-frequency-annotations) | cg69
[ClinVar](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#clinvar-annotations) | clinvar_20160302
[COSMIC](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#cosmic-annotations) | cosmic70
[ICGC](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#icgc-annotations) | icgc21
[NCI60](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#nci60-annotations) | nci60

As we will see, filter-based annotation is much the same as region- and gene-based. However, when downloading for the first time we have to add an additional `-webfrom annovar` argument into the command as the filter-based files are provided by the annovar website. For the purposes of this session, all the files have already been provided.

```{}
../software/annovar/annotate_variation.pl -buildver hg19 -downdb -webfrom annovar cosmic70 ../software/annovar/humandb/
```

It seems natural to use the latest version of a database for the most-comprehensive annotation. However, the annovar website suggests that we be cautious with regards to [dbSNP](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#dbsnp-annotations). As dbSNP has been expanded to include variants from other large-scale projects (such as 1000 genomes) it has become somewhat "contaminated" by variants with very low frequency or with clinical relevance. Thus, ***dbSNP129*** is regarded as being the last clean version of the database.


Now we select the `filter` option in annovar with a `dbtype` of `snp129`
```{}
../software/annovar/annotate_variation.pl -filter -buildver hg19 -dbtype snp129  HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/
```

This time we get two files created; `HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_snp129_dropped` and `HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_snp129_filtered`. The difference being that the `.._dropped` file contains details of all your variants that ***are*** in the database

```{}
head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_snp129_dropped
```

```{r echo=FALSE,comment=NA}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_snp129_dropped",intern=TRUE),sep="\n")
```

Thus, the `.._filtered` file contains all the variants ***not found*** in the database. The implication being that we want to remove common variants found amongst healthy individuals from our list. There is nothing to stop us using multiple versions of dbSNP in the filtering as we will see later-on.

We can perform annotation against 1000 genomes dataset with the following:-

```{}
../software/annovar/annotate_variation.pl -filter -dbtype 1000g2015aug_all -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/
```

As before, `..._dropped` and `...filtered` files are produce. Again the `.._dropped` file contains variants that were found in the database. However, rather than printing and ID for the variant, it prints the minor allele frequency that is observed for that variant.

```{}
head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_EUR.sites.2014_10_dropped
```

```{r echo=FALSE,comment=NA}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_EUR.sites.2014_10_dropped",intern=TRUE),sep="\n")
```

We can also provide a minor allele frequency (maf) threshold. This will mean that only variants with a maf greater than the threshold will be printed to the `..._dropped` file.

```{}
../software/annovar/annotate_variation.pl -filter -dbtype 1000g2015aug_all -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/ -maf 0.05
```

The commands above annotated against variants derived from european indivduals. If we have individuals of a different ethnicity, we probably want to use a different cohort. In this case, we annotate against frequencies observed in european individuals.

```{}
../software/annovar/annotate_variation.pl -filter -dbtype 1000g2014oct_eur -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/
```

## Healthy or disease individuals?

How you choose to interpret the filtering results depends on whether you are interested in variants that are exclusive to healthy or diseased individuals.

Both dbSNP and 1000 genomes report frequencies amongst healthy individuals. The "complete genomics" (`cg69`) set also provides variant frequencies for 69 healthy subjects (including 46 unrelated subjects). Particularly useful for those with exome data is the ESP (Exome Sequencing Project) which aims to identify variants in exonic regions from over 6000 individuals; including however some subjects with disease. 

```{}
../software/annovar/annotate_variation.pl -filter -dbtype clinvar_20160302 -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/
```

`annovar` can also annotate against variants that are relevant to diseases such as cancer. For example, there is the ***COSMIC*** (Catalogue of Somatic Mutations in Cancer).

```{}
../software/annovar/annotate_variation.pl -filter -dbtype cosmic70 -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/
```

Again, the variants found in the database are printed in the `_dropped` file along with details of the mutation in **COSMIC** that they match.

```{r echo=FALSE,comment=NA}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_cosmic70_dropped",intern=TRUE),sep="\n")
```

This is slightly contrary to the 1000 genomes and dbSNP filtering we did previously where we wanted to retain variants that were not found in a database. Fortunately in the next section we will look at any approach where we can get results for all our variants in a single table that we can develop our own filtering criteria for.

### LJB* (dbNSFP) non-synonymous variants annotation


## Combining multiple annotations and filters

Up to this point, we have amassed quite a collection of tables of our set of variants annotated / filtered against various databases. A particularly appealing feature of `annovar` is that we can define protocols to combine multiple filters and collect the results in the same table. 

The `-protocol` argument is used to list what types of annotation / filtering you want to be applied. With the `-operation` argument we have to specify whether each annotation is a gene (`g`), region (`r`) or filter (`f`) -based annotation. You need to take care to provide the same number of names in both the `-protocol` and `-operation` arguments.

```{}
../software/annovar/table_annovar.pl HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/ -buildver hg19 -out myanno -remove -protocol refGene,cytoBand,gwasCatalog,genomicSuperDups,dgvMerged,snp129,1000g2015aug_all,esp6500si_all,clinvar_20160302,cosmic70,nci60 -operation g,r,r,r,r,f,f,f,f,f,f -nastring NA -csvout
```

The file `myanno.hg19_multianno.csv` is created. By specifying the `-remove` argument in the command, any tables that are created as part of the process are removed. This `.csv` file can now be viewed in Excel, or LibreOffice in our case (right-click on the file)

Having opened this file into a spreadsheet program we can of course filter and sort the data as we see fit. However, due to the sheer number of columns we have generated this may prove a bit unwieldy. Therefore we are going to introduce an R package that will make the interrogation and comprehension of such a table more intuitive.

# Using R to interrogate and filter an annotation table

## Introducing dplyr

`dplyr` is a very powerful package for manipulating data in tables and is part of the *tidy-verse* collection of packages that provide a seamless and consistent interface to exploring and analysing data.

![](images/data-cycle.png)

The process of analysing a dataset is rarely a linear one, and will generate scores of intermediate plots and tables along the way that don't make their way to a final publication. Packages such as `dplyr` and `ggplot2` (for plotting) provide an interface whereby the effort to produce meaningful summaries and graphics of our data is minimised. This allows us to worry more about the interpretation and what questions we should ask, rather than writing code.





The [cheatsheet](https://www.rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf) is highly recommended and summarises all the functions within the package. We won't have time to cover everything here today.

For more information on `dplyr` see our one-day course [Data Manipulation and Visualisation using R](http://bioinformatics-core-shared-training.github.io/r-intermediate/)

The first step is to load the `dplyr` library.
```{r message=FALSE}
library(dplyr)
```

An incredibly useful but simple function is `tbl_df` which converts the standard R `data.frame` into a special `dplyr` one that has a nicer standard display. Rather than filling the whole R console we can see a snapshot of the data frame and summaries of the data contained in each column


```{r}
anno <- read.csv("myanno.hg19_multianno.csv")
anno <- tbl_df(anno)
anno
```

`dplyr` introduces a series of verbs, each of which is capable of performing a specific operation on a data frame in an efficient manner. The syntax of the verbs is very consistent. Not only does this allow us to easily switch between using different operations, but we can chain the operations together as we will see later.

### select

`select` is the first of the `dplyr` verbs that can be used to pick certain columns from the table

```{r}
select(anno, Func.refGene)
```

No great advance in our capabilities yet, as we already know how to select a variable from a table using the `$` operator

```{r eval=FALSE}
anno$Func.refGene
```

However, the syntax to select multiple columns is a lot more concise. Mutiple columns can be selected by separating their names with a comma

Compare:- 
```{r}
select(anno, Func.refGene,ExonicFunc.refGene)
```

to

```{r}
anno[,c("Func.refGene","ExonicFunc.refGene")]
```

Non R users probably find these commands a bit obtuse
- why the `[ ]`?
- what is `c`?
- need to remember the row and column index
- `[,...]` means display all rows

However, `select` is a lot more intuitive; easier to convert to the sentance “Select the Func.refGene and ExonicFunc.refGene columns from patients”

We can select a range of columns with a `:`

```{r}
select(anno, cytoBand:X1000g2015aug_all)
```

Or suppress columns by putting with a `-` in front of their name

```{r}
select(anno, -End)
```

Combinations of the above are possible

```{r}
select(anno, cytoBand:X1000g2015aug_all,-snp129,-AAChange.refGene)
```

Some convenient functions are available so we don't need to specify the column names in full

```{r}
select(anno, starts_with("CLN"))
select(anno, contains("1000g"))
```

*******
Exercise

- Select the columns `Func.refGene` to `AAChange.refGene` 
    + you could try the convenience functions `contains` `ends_with` to do this, or using the `:` notation
- Now remove the column `GeneDetail.refGene`
- Append the chromosome start and end position columns

```{r echo=FALSE}
select(anno, Chr:End, contains("refGene"),-GeneDetail.refGene)
```

******

We should note at this point, that we have not actually changed the data frame at any point; just created different views of the data. 

```{r}
anno
```



### filter

A particularly useful operation for us is going to be `filter`. Again, there is nothing here that we couldn't do with "base R", but it will allow us to create more readable code.

The syntax is consistent with `select` and uses the usual selection techniques in R such as `==`.

Selecting variants on chromsome 1

```{r}
filter(anno, Chr ==1)
```

Selecting *exonic* variants
```{r}
filter(anno, Func.refGene == "exonic")
```

Multiple conditions can be separated by a `,`

"*Exonic variants on chromosome 1 that are no synonymous or unknown*"
```{r}
filter(anno, Func.refGene == "exonic", Chr == 1, ExonicFunc.refGene != "synonymous SNV", ExonicFunc.refGene != "unknown")
```

Variants that are found in dbSNP
```{r}
filter(anno, is.na(snp129))
```

Variants that are found in GWAS

```{r}
filter(anno, !is.na(gwasCatalog))
```



*******
Exercise

- Find which variants are present COSMIC database
- Find which variants are not in segmental duplications
- Find all variants that are in TP53 and exonic
- Find all variants that are in TP53, exonic and less than 0.05 maf in 1000 genomes individuals and less than 0.05 frequency in the exon sequencing project

*******

## Combining commands with "pipes"

A slighty annoyance with the first exercise above is that once we've filtered for COSMIC variants we don't actually see the COSMIC column appearing in the output:-

```{r}
filter(anno, !is.na(cosmic70))
```

Given what we have done so far, it would seem natural to follow this `filter` statement with a `select` to view the columns we are interested in

It is quite common to nest commands in R into a single line

```{r}
anno <- tbl_df(read.csv("myanno.hg19_multianno.csv"))
```

which we read as 

> apply the function `tbl_df` to the result of reading the file myanno.hg19_multianno.csv

The following line of R code is perfectly valid and does exactly what we want. However, reading the code back to ourselves is a bit clunky.

```{r}
select(filter(anno, !is.na(cosmic70)), Chr:Alt, cosmic70)
```

> select from the result of filter COSMIC variants, the columns Chr to Alt and cosmic70

or we can split over multiple lines
```{r}
select(filter(anno, !is.na(cosmic70)),
       Chr:Alt, cosmic70)
```


We always have to work out what the first statement was, and work forwards from that.
 - the arguments to a function are far-away from the function call


***piping*** is a familiar concept in unix where the output of one command is used as input to the following line. The `dplyr` package allows such operations to be performed in R via use of the ` %>% ` operation from the `magrittr` package

- Shortcut in RStudio: CTRL + SHIFT + M, CMD + SHIFT + M (Mac)

`dplyr` functions are designed to take a data frame as their first argument, and return a data frame as an output. This makes them ideal to use with a pipe.

The code from before can be rewritten as follows, making it easier to see the order in which the operations were performed.

```{r}
filter(anno, !is.na(cosmic70)) %>% 
  select(Chr:Alt, cosmic70)
```


*******
Exercise

- Explain in words what the following chain of commands is doing

```{r}
filter(anno, Func.refGene=="exonic") %>% 
  filter(!is.na(genomicSuperDups)) %>% 
  filter(is.na(snp129)) %>% 
  filter(X1000g2015aug_all < 0.05) %>% 
  select(Chr:ExonicFunc.refGene,-GeneDetail.refGene,-Func.refGene,X1000g2015aug_all)
```


*******
Exercise

- Find which variants are present in the GWAS catalog
- Print the chromsome, start, end, and Gene name for these variants
- Use a "pipe" command to write your answer

```{r}
filter(anno, !is.na(gwasCatalog)) %>% 
  select(Chr:End, Gene.refGene, gwasCatalog)
```

*******

### arrange

The `arrange` function can be used to order rows. For example ordering according to 1000 genomes frequency

```{r}
arrange(anno, X1000g2015aug_all)
```

However, we can't actually see the column, so using a `select` statement afterwards seems a good idea

```{r}
  arrange(anno, X1000g2015aug_all) %>% 
  select(Chr:End, X1000g2015aug_all)
```

We can use the convenient function `desc` to switch to descending order

```{r}
  arrange(anno, desc(X1000g2015aug_all)) %>% 
  select(Chr:End, X1000g2015aug_all)
```


*******
Exercise

- Which genes have the greatest number of exonic variants?

*******


```{r echo=FALSE}
filter(anno, Func.refGene=="exonic") %>% 
  select(Chr:Gene.refGene) %>% 
  count(Gene.refGene) %>% 
  arrange(desc(n))
```



# Using annovar to make prediction


```{}
../software/annovar/table_annovar.pl HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/ -buildver hg19 -out myanno.pred -remove -protocol dbnsfp30a -operation f -nastring NA -csvout
```


```{r}
preds <- read.csv("myanno.pred.hg19_multianno.csv") %>% 
  tbl_df

preds
```

