---
title: "Annotating and Filtering"
author: "Mark Dunning"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
---

# Outline

Imagine that we've called a set of somatic variants from our paired tumour and normal sample.

- Using `annovar` to generate a rich set of annotations for our variants
- Recent advances in R ("`dplyr`") that allow us to filter and subset our data
- How we can prioritise our results, and write pipelines for doing so

# Working Directory

```{}
/home/participant/Course_Materials/Day2/
```


## Why annotate

- Can have huge list of variants after SNV-calling
    + order of millions even after filtering
- Not all will have consequences
- Need to functionally annotate biological and biochemical function
- Determine which variants have been previously-identified
    + in public databases
    + amongst healthy / diseased individuals
    
## Introduction to annovar

There are many annotation resources available, but we have chosen to demonstrate `annovar` for ease of usage and compatibility across platforms. `annovar` is a suite of perl scripts that can perform a variety of annotation and filtering tasks. It can also collate the results from annotating against different databases into one convenient table.

***Important:- annovar license***

`annovar` is free for academic use. However, you need to supply a valid academic email address in order to download the software. If you wish to use `annovar` after the course. You will have to fill out the form on the [annovar](http://annovar.openbioinformatics.org/en/latest/user-guide/download/) website. 


`annovar` provides "gene-based", region-based" and "filter-based" annotation and relies on *pre-downloaded* annotation files so all queries are done offline. Files can be downloaded from the UCSC genome browser or from the annovar website. Some of these files can be quite large, but fortunately we only have to download them once. For the course, we have downloaded some of these files to the directory `/home/participant/Course_Materials/software/annovar/humandb`. Each annotation file has a genome version associated with it. For our purposes we are going to be using `hg19`. 

For your reference the script to download all the resources used in this session is

`/home/participant/Course_Materials/software/annovar/annovar_commands.sh`

```{}
ls -l ../software/annovar/humandb
```


## Preparing files for annovar

Some of the commands in `annovar` require a specific input format. Fortunately, they have provided a script that can convert from popular file formats to their own format. 

We convert our `.vcf` file using a script from annovar `convert2annovar.pl`

```{}
../software/annovar/convert2annovar.pl -format vcf4old -includeinfo HCC1143_vs_HCC1143_BL.flagged.muts.vcf > HCC1143_vs_HCC1143_BL.flagged.muts.avinput
```

```{}
-format: vcfold, the input is in vcf format
-includeinfo: Include the INFO fields from the vcf file
>: redirect to the file of our choosing
```


```{r echo=FALSE,comment=NA} 
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput",intern=TRUE),sep="\n")
```

The first five columns are compulsory have to contain

- Chromosome
- Start Position
- End Position
- Reference base(s)
- Observed base(s)

All other columns are optional. They are not used in the annotation process, but will get carried-along with any annotation and filtering we perform

### Region-based Annotation

Gene-based annotation is the default operation in `annovar` and is used to identify variants that cause protein-coding changes. Various gene-definition systems are supported.

Database  | Keyword
------------- | -------------
RefSeq  | refGene
UCSC    | knownGene
Ensembl | ensGene
Gencode | wgEncodeGencodeBasicV19



The following command will annotate the file we have just prepared against the reference file located in `../software/annovar/humandb/hg19_refGene.txt`.


```{}
../software/annovar/annotate_variation.pl -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb
```

Two files are created:

-`HCC1143_vs_HCC1143_BL.flagged.muts.avinput.variant_function`

This file lists ***all*** variants in the input file are pre-pends two columns. The first states whether a given variant is within an exonic or not and gives the relevant gene name. 

```{r echo=FALSE,comment=NA}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.variant_function",intern=TRUE),sep="\n")
```


- `HCC1143_vs_HCC1143_BL.flagged.muts.avinput.exonic_variant_function` 

This file lists for the ***coding variants only*** whether the stated base-change will lead to a change in protein or not. Full definitions of all the terms is on the [annovar website](http://annovar.openbioinformatics.org/en/latest/user-guide/gene/)

```{r echo=FALSE,comment=NA}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.exonic_variant_function",intern=TRUE),sep="\n")
```



### Region-based annotation

More comprehensive details on the types of region-based filtering that is supported can be found on the [http://annovar.openbioinformatics.org/en/latest/user-guide/region/](annovar website). Here we list some examples along with the "keyword" to be used when downloading and annotating against the database


Database  | Keyword 
------------- | -------------
Cytoband  | cytoband 
microRNA and snoRNA | wgRna 
Conserved Regions  | phastConsElements46way
Transcription Factor Binding Sites | tfbsConsSites
Published GWAS | gwasCatalog
Database of Genomic Variants | dgvMerged
Segmental Duplications | genomicSuperDups

or region definitions of your choosing in `.bed` or `.gff` format. Annotation against regions identified in ENCODE project is also described in its own section on the [annovar website](http://annovar.openbioinformatics.org/en/latest/user-guide/region/#identify-variants-in-encode-annotated-regions)

The simplest form of region annotation is probably when we want to look at which "cytoband" each variant belongs to. This uses a reference file that contains the coordinates for each cytoband in a given genome version. 

```{}
head ../software/annovar/humandb/hg19_cytoBand.txt
```

```{r echo=FALSE,comment=NA}
cat(system("head ../software/annovar/humandb/hg19_cytoBand.txt",intern=TRUE),sep="\n")
```

If didn't already have such a file it could be downloaded using the following. The `cytoBand` keyword (from the table above) is used along with the build version `hg19`.

```{}
../software/annovar/annotate_variation.pl -buildver hg19 -downdb cytoBand ../software/annovar/humandb/
```

To perform the annotation, we use need to specify `-regionanno`, to perform region-annotation and set the `dbtype` to `cytoBand`. 

```{}
../software/annovar/annotate_variation.pl -regionanno -build hg19 -dbtype cytoBand  HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/
``` 

This should result in a file being created with `.cytoBand` added as a file extension to your input file (e.g. `HCC1143_vs_HCC1143_BL.flagged.muts.avinput.cytoBand`). An extra column has been added to indicate which chromosome band each variant is located on. In the event of a variant spanning multiple bands (obviously not in this case), the multiple bands would be listed here.

```{}
head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.cytoBand
```

```{r echo=FALSE,comment=NA}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_cytoBand",intern=TRUE),sep="\n")
```

In another example, we could annotate against locations that have been reported in previous GWAS studies. The command is the same as the previous example, except we substitute `cytoBand` for the keyword for GWAS annotation (`gwasCatalog`)

```{}
../software/annovar/annotate_variation.pl -regionanno -build hg19 -dbtype gwasCatalog HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb
```

```{r echo=FALSE,comment=NA}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_gwasCatalog",intern=TRUE),sep="\n")
```

*******
Exercise

- Select any of the databases from the table that you think might be relevant to your study, and annotate the example SNV calls
    + all the database files should already be downloaded.
    
*******




### Filter-based annotation

Unlike the annotation performed in the previous section, which was based purely on the genomic location, *filter-based* annotation also takes the base-change into account. One can also use the observed frequency in the database population as a means to filter.

As before, many types of database are available. In this table, we list the *latest* version available:- 

Database  | Keyword 
------------- | -------------
dbSNP  | snp138
1000 Genomes | 1000g2015aug
Exome Aggregation Consortium | exac03
Exome Sequencing Project | esp6500si_all 
Complete Genomics Individuals | cg69
ClinVar | clinvar_20160302
COSMIC | cosmic70
ICGC | icgc21
NCI60 | nci60

- Which are disease and healthy?
- dbSNP versions

For example, if we want to see which of our variants are present a in database of known (or common) SNPs we could use the `dbSNP` database. 

Now we select the `filter` option in annovar with a `dbtype` of `snp138`
```{}
../software/annovar/annotate_variation.pl -filter -buildver hg19 -dbtype snp138  HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/
```

This time we get two files created; `HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_snp129_dropped` and `HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_snp129_filtered`. The difference being that the `.._dropped` file contains details of all your variants that ***are*** in the database

```{}
head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_snp129_dropped
```

```{r echo=FALSE,comment=NA}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_snp129_dropped",intern=TRUE),sep="\n")
```

Thus, the `.._filtered` file contains all the variants ***not found*** in the database. 

We can perform annotation against 1000 genomes variants in europeans with the following:-

```{}
../software/annovar/annotate_variation.pl -filter -dbtype 1000g2014oct_eur -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/
```

As before, `..._dropped` and `...filtered` files are produce. Again the `.._dropped` file contains variants that were found in the database. However, rather than printing and ID for the variant, it prints the minor allele frequency that is observed for that variant.

```{}
head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_EUR.sites.2014_10_dropped
```

```{r echo=FALSE,comment=NA}
cat(system("head HCC1143_vs_HCC1143_BL.flagged.muts.avinput.hg19_EUR.sites.2014_10_dropped",intern=TRUE),sep="\n")
```

We can also provide a minor allele frequency (maf) threshold. This will mean that only variants with a maf greater than the threshold will be printed to the `..._dropped` file.

```{}
../software/annovar/annotate_variation.pl -filter -dbtype 1000g2014oct_eur -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/ -maf 0.05
```

The commands above annotated against variants derived from european indivduals. If we have individuals of a different ethnicity, we probably want to use a different cohort. In this case, we annotate against frequencies observed in asian individuals.

```{}
../software/annovar/annotate_variation.pl -filter -dbtype 1000g2014oct_afr -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/
```

Clinical variants.....

```{}
../software/annovar/annotate_variation.pl -filter -dbtype clinvar_20160302 -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/
```


Exome Aggreation Consortium (ExAC)

```{}
../software/annovar/annotate_variation.pl -filter -dbtype exac03 -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/
```

Complete Genomics

```{}
../software/annovar/annotate_variation.pl -filter -dbtype cg69 -buildver hg19 HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/
```


### LJB* (dbNSFP) non-synonymous variants annotation


## Combining multiple annotations and filters

Up to this point, we have amassed quite a collection of tables of our set of variants annotated / filtered against various databases. A particularly appealing feature of `annovar` is that we can define protocols to combine multiple filters and collect the results in the same table. 

The `-protocol` argument is used to list what types of annotation / filtering you want to be applied. With the `-operation` argument we have to specify whether each annotation is a gene (`g`), region (`r`) or filter (`f`) -based annotation. You need to take care to provide the same number of names in both the `-protocol` and `-operation` arguments.

```{}
../software/annovar/table_annovar.pl HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/ -buildver hg19 -out myanno -remove -protocol refGene,cytoBand,gwasCatalog,snp129,1000g2014oct_eur,1000g2014oct_afr,exac03,clinvar_20160302 -operation g,r,r,f,f,f,f,f -nastring NA -csvout
```

The file `myanno.hg19_multianno.csv` is created. By specifying the `-remove` argument in the command, any tables that are created as part of the process are removed. This `.csv` file can now be viewed in Excel, or LibreOffice in our case (right-click on the file)

```{}
../software/annovar/table_annovar.pl HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/ -buildver hg19 -out myanno -remove -protocol refGene,exac03 -operation g,f -nastring NA -csvout
```


# Using R to interrogate and filter an annotation table

## Introducing dplyr

`dplyr` is a very powerful package for manipulating data in tables. The [cheatsheet](https://www.rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf) is highly recommended and summarises all the functions within the package. We won't have time to cover everything here today.

For more information on `dplyr` see our one-day course [Data Manipulation and Visualisation using R](http://bioinformatics-core-shared-training.github.io/r-intermediate/)


```{r message=FALSE}
library(dplyr)
```

the `tbl_df` function is used to convert the standard R data.frame into a special `dplyr` one that has a nicer standard display.

```{r}
anno <- read.csv("myanno.hg19_multianno.csv")
anno <- tbl_df(anno)
anno
```

### select

`select` is the first of the `dplyr` verbs that can be used to pick certain columns from the table

```{r}
select(anno, Func.refGene)
```

Mutiple columns can be selected by separating their names with a comma

```{r}
select(anno, Func.refGene,ExonicFunc.refGene)
```

We can select a range of columns with a `:`

```{r}
select(anno, cytoBand:X1000g2014oct_afr)
```

Or suppress columns with a `-` in front of their name


```{r}
select(anno, -End)
```

Combinations of the above are possible

```{r}
select(anno, cytoBand:X1000g2014oct_afr,-snp129)
```

```{r}
select(anno, starts_with("CLN"))
```



*******
Exercise

- Select the columns `Func.refGene` to `AAChange.refGene` 
    + you could try the convenience function `contains` to do this, or using the `:` notation
- Now remove the column `GeneDetail.refGene`
- Append the chromosome start and end position columns

```{r echo=FALSE}
select(anno, Chr:End, contains("refGene"),-GeneDetail.refGene)
```

******

We should note at this point, that we have not actually changed the data frame at any point; just created different views of the data. 

```{r}
anno
```

### mutate

Sometimes we might want to change the data available in the table, either by adding new columns or modifying existing ones. This is supported by the `mutate` function. We have to specify the column name (which doesn't already have to exist) and some new values to put in this column. Any valid R function can be used to create values, provided the number of values to try add is the same as the number of rows.

For example, we might want to change the values stored for `Chr` so they have the usual prefix.

```{r}
mutate(anno, Chr=paste0("Chr", Chr))
```

### filter

A particularly useful operation for us is going to be `filter`. The syntax is consistent with `select` and uses the usual selection techniques in R such as `==` 

```{r}
filter(anno, Chr ==1)
```

```{r}
filter(anno, X1000g2014oct_eur < 0.05)
```


Again, multiple conditions can be added

```{r}
filter(anno, Chr==1, Func.refGene == "exonic")
```

```{r}
filter(anno, Func.refGene != "intronic", Func.refGene != "exonic")
```

A useful shortcut function is `count`, which will give a table of how many values are found in a particular column

```{r}
count(anno, Func.refGene)
```


*******
Exercise

- Find all variants that are in TP53 and exonic
- Find all variants that are in TP53, exonic and less than 0.05 maf in european and african 1000 genomes individuals
- Find which variants are present in the GWAS catalog

*******

### Combining commands with "pipes"

piping is a familiar concept in unix where the output of one command is used as input to the following line. The `dplyr` package allows such operations to be performed in R via use of the ` %>% ` operation from the `magrittr` package

`dplyr` functions are designed to take a data frame as their first argument, and return a data frame as an output.

Say we want all exonic variants and then display all the columns that relate to the gene-based annotation

```{r}
filter(anno, Func.refGene=="exonic") %>% 
  select(contains("refGene"))
```


*******
Exercise

- Find which variants are present in the GWAS catalog
- Print the chromsome, start, end, and Gene name for these variants

```{r}
filter(anno, !is.na(gwasCatalog)) %>% 
  select(Chr:End, Gene.refGene, gwasCatalog)
```

*******

### arrange

The `arrange` function can be used to order rows. For example ordering according to 1000 genomes frequency

```{r}
arrange(anno, X1000g2014oct_eur)
```

However, we can't actually see the column, so using a `select` statement afterwards seems a good idea

```{r}
  arrange(anno, X1000g2014oct_eur) %>% 
  select(Chr:End, X1000g2014oct_eur)
```

We can use the convenient function `desc` to switch to descending order

```{r}
  arrange(anno, desc(X1000g2014oct_eur)) %>% 
  select(Chr:End, X1000g2014oct_eur)
```


*******
Exercise

- Which genes have the greatest number of exonic variants?

*******


```{r echo=FALSE}
filter(anno, Func.refGene=="exonic") %>% 
  select(Chr:Gene.refGene) %>% 
  count(Gene.refGene) %>% 
  arrange(desc(n))
```


## Joining tables

```{r}
library(VariantAnnotation)
vcf <- readVcf("HCC1143_vs_HCC1143_BL.flagged.muts.vcf","hg19")
vcfInfo <- as.data.frame(info(vcf))
bind_cols(anno,vcfInfo)

```


```{r}
vcfDf <- data.frame(Chr=seqnames(vcf),Start=start(vcf),End=end(vcf),vcfInfo, GT = geno(vcf)$GT,PM=geno(vcf)$PM)
anno <- left_join(anno, vcfDf,by=c("Chr","Start","End"))
write.csv(anno, file="myanno.hg19_multianno.joined.csv",row.names = FALSE)

```

# Using annovar to make prediction


```{}
../software/annovar/table_annovar.pl HCC1143_vs_HCC1143_BL.flagged.muts.avinput ../software/annovar/humandb/ -buildver hg19 -out myanno.pred -remove -protocol dbnsfp30a -operation f -nastring NA -csvout
```


```{r}
preds <- read.csv("myanno.pred.hg19_multianno.csv") %>% 
  tbl_df

preds
```

